[Networking] FLIP: Message Forensic (MF) System (#195)
onflow/flips: protocol/20230908-message-forensic.md

| Status        | Proposed                                            |
:-------------- |:----------------------------------------------------|

> | **FLIP #**    | [195](https://github.com/onflow/flips/pull/195) |

Author: ```suggestion
| **FLIP #**    | [259](https://github.com/onflow/flips/issues/259) |
```
@kc1116

Author: @KshitijChaudhary666 as @AlexHentschel mentioned, I have moved out of Flow, and this FLIP has been handed off to @kc1116. I followed your guidelines on creating an issue and updating the FLIP number. However, as I am not part of the Flow org, I cannot amend the PR.

Reviewer: thank you for making the change Yahya! I will commit the change

> | **FLIP #**    | [4697](https://github.com/onflow/flow-go/pull/4697) |

Reviewer 2: ```suggestion
| **FLIP #**    | [195](https://github.com/onflow/flips/pull/195) |
```

| **Updated**   | 8 September 2023                                    |

# Message Forensic (MF) System

## Summary
This FLIP discusses and compares two potential solutions for the Message Forensic (MF) system in the Flow protocol 
— a system that enables Flow protocol to attribute protocol violations to the original malicious sender. 
The two solutions under consideration are: (1) GossipSub Message Forensic (GMF), and (2) Enforced Flow-level Signing Policy For All Messages. 
We delve into both, listing their pros and cons, to determine which would be more feasible given the considerations of ease of implementation, 
performance efficiency, and security guarantees.

Our analysis finds the "Enforced Flow-level Signing Policy For All Messages" to be the more promising option, 
offering a generalized solution that doesn’t hinge on a specific external protocol to send the message (e.g., GossipSub),
steering clear of the complexities tied to maintaining GossipSub envelopes and dodging the necessity of duplicating GossipSub 
router’s signature verification procedure at the engine level. Furthermore, it meshes well with the Flow protocol’s existing state.

## Review Guide

> This FLIP is presented as a Pull Request (PR) in the `flow-go` repository. We welcome reviewers to express their opinions and share feedback directly on the PR page, aiming for a structured and productive discussion. To aid this, please adhere to one of the following response frameworks:

Reviewer 2: The linked PR is closed. Should this be updated to ask for feedback on the current PR ([195](https://github.com/onflow/flips/pull/195)) instead?

Author: https://github.com/onflow/flips/pull/195/commits/b44a9db5495e54191392bc9a862be2b8d931c4a0

1. I favor the "Flow-level Signing Policy (FSP)" and here are my thoughts:
2. I support the "GossipSub Message Forensic (GMF)" approach, articulating my views as follows.
3. I find both propositions unsatisfactory, elucidating my stance with.

## Problem Overview
Within the Flow protocol, nodes converse through a networking layer, a medium which undertakes the dispatch and receipt of messages among nodes
via different communication methods: unicast, multicast, and publish. Unicast messages are transmitted to a single recipient over a direct connection.
The multicast and publish, on the other hand, utilize a pub-sub network constructed with the LibP2P GossipSub protocol for message 

> dissemination. However, this protocol encounters challenges in message attribution, particularly in determining the 

Reviewer 3: The context here is great. It would be also clearer if we recall the Flow protocol state assumptions (for each node, the protocol state tracks 2 public keys, a staking and a networking key)

initial sender of a message, which is critical for identifying protocol violations and penalizing the malicious initiators. 
Below, we break down the issues seen in different communication strategies and their implications.

### Single Recipient (Unicast) Message Attribution
Unicast messages have an implicit attribution to the sender, i.e., they are attributed to the remote node of the connection
on which the message is received. Albeit, the situation differs when the unicast message carries the explicit signature of the
sender or other attribution data (e.g., the signature of the original sender in case of a forwarded unicast).
In such cases, the message is attributed to the sender that is specified in the attribution data. Nevertheless, the current state of
Flow protocol does not enforce the attribution data to be present in the unicast messages. 

> A prime example is the [`ChunkDataResponse` messages](https://github.com/onflow/flow-go/blob/master/engine/execution/provider/engine.go#L326-L337) sent by the Execution Nodes to the Verification Nodes over a unicast. 

Reviewer 2: ```suggestion
A prime example is the [`ChunkDataResponse` messages](https://github.com/onflow/flow-go/blob/a85341d949c8a4e42e3d1d62bcf6fbf3d7f76bd9/engine/execution/provider/engine.go#L326-L337) sent by the Execution Nodes to the Verification Nodes over a unicast. 
```

Hence a node in Flow protocol cannot prove that a received unicast message is sent by a specific node, unless the message carries the explicit attribution data.

### Group and All-Node (Multicast and Publish) Message Attribution

> When dealing with multicast or publish messages, the GossipSub router initially [signs it](https://github.com/libp2p/go-libp2p-pubsub/blob/master/sign.go#L109-L134) with a local node's networking key 

Reviewer 2: Suggestion: Use permalink instead of link to `master`, in case file gets updated.

Author: https://github.com/onflow/flips/pull/195/commits/544c59f3de4fa83982b35dc40a8263f46f0e1ba3

Hence, on the receiving path, the networking layer performs two rounds of decoding before delivering the event to the Flow-protocol engines. 
- For the unicast cases, the networking layer reads the bytes from the stream, and decodes the `messages.Message` protobuf type from the bytes. Then, the `messages.Message.Payload`
is decoded to an `interface{}` type event using CBOR, and is delivered to the Flow-protocol engines.
- For the pubsub cases, the networking layer receives the `pb.Message` type from GossipSub, and decodes the `messages.Message` protobuf type from the `pb.Message.Data` bytes. 
Then, the `messages.Message.Payload` is decoded to an `interface{}` type event using CBOR, and is delivered to the Flow-protocol engines.

In the rest of this FLIP, we refer to `pb.Message` as the GossipSub envelop, and to the `message.Message` as the Flow message. Hence, in a nutshell:
- For the unicast case, the event is wrapped in a Flow message type.
- For the pubsub case, the event is wrapped in a Flow message type, which is then wrapped in a GossipSub envelope type.

> - The GossipSub envelop contains the signature of the sender over the entire envelope. The signature is generated using the networking key of the sender by the GossipSub router.

Reviewer 2: ```suggestion
- The GossipSub envelope contains the signature of the sender over the entire envelope. The signature is generated using the networking key of the sender by the GossipSub router.
```

## Proposal-1: GossipSub Message Forensic (GMF)
As the first option to (partially) mitigate the gap in the network to hold the malicious senders accountable,
we introduce the GossipSub Message Forensic (GMF) mechanism designed to integrate GossipSub authentication data seamlessly with the 
Flow protocol engines. The principal aim is to enhance message authenticity verification through multicast and publish,
focusing on origin identification and event association. Here, we elucidate the method, its interface, and delve into the operational specifics, 
weighing its pros and cons against potential alternatives.
In this proposal, we propose a mechanism to share the GossipSub authentication data with the Flow protocol. We call this mechanism
GossipSub Message Forensic (GMF). 

> At the current state of the code, the networking layer receives the [raw GossipSub envelop](https://github.com/onflow/flow-go/blob/master/network/p2p/p2pnet/internal/readSubscription.go#L48) from its underlying pubsub router,

Reviewer 2: ```suggestion
At the current state of the code, the networking layer receives the [raw GossipSub envelope](https://github.com/onflow/flow-go/blob/master/network/p2p/p2pnet/internal/readSubscription.go#L48) from its underlying pubsub router, 
```

validates it, and passes the incorporated Flow message to for further decoding. The Flow message is then decoded to an `interface{}` type event,
and is delivered to the Flow protocol engines. The idea is to pass along the GossipSub envelop to the Flow protocol engines together with their `interface{}` type event.
In this way, the Flow protocol engines can have access to the GossipSub authentication data. The rough idea is depicted in the following figure. 
Note that it is not a production-ready code, we need to architect the solution in an Object Oriented Manner with proper abstraction and encapsulation of the interface. In this figure,
we intentionally pass along the envelope together with the event for sake of simplicity. In the actual implementation, we need to abstract the envelope and event in a single object 

> and pass it to the Flow protocol engines. We have to also account for the unicast case that does not have a GossipSub envelop.

Reviewer 2: ```suggestion
and pass it to the Flow protocol engines. We have to also account for the unicast case that does not have a GossipSub envelope.
```

by two new methods: `VerifyGossipSubMessage`. The `VerifyGossipSubMessage` method takes an origin identifier, an event `interface{} type`,
as well as the GossipSub message as input, and verifies the signature of the message against the networking public key of the origin id. 
The method returns true if the signature is valid, and false otherwise.
```go

type EngineRegistry interface {
	// VerifyGossipSubMessage takes a Flow identifier as well as a GossipSub message as input, and verifies the signature of the
	// message. The method returns true if the signature is valid, and false otherwise.
	// Args:
	//  - originId: The Flow identifier of the node that originally sent the message.
	// - message: The GossipSub message that is associated with the event.
	// Returns:
	//  - bool: True if the signature is valid, and false otherwise.
	//  - error: An error if the message is not found.
	VerifyGossipSubMessage(originId flow.Identifier, event interface{}, message *pb.Message) (bool, error)

	// Other methods are omitted for brevity
	// ...
}
```

### Implementation Complexities

> - This proposal requires refactoring the message processing pipeline of the networking layer to pass along the GossipSub envelop together with the event to the Flow protocol engines. We need to

Reviewer 2: ```suggestion
- This proposal requires refactoring the message processing pipeline of the networking layer to pass along the GossipSub envelope together with the event to the Flow protocol engines. We need to
```

>   also refactor the `Engine`'s `Process` method in a way that it can accept the GossipSub envelop together with the event. The implementation should account for proper encapsulation and abstraction of the interface, 

Reviewer 2: ```suggestion
  also refactor the `Engine`'s `Process` method in a way that it can accept the GossipSub envelope together with the event. The implementation should account for proper encapsulation and abstraction of the interface,
```

>   to also support the unicast case that does not have a GossipSub envelop.

Reviewer 2: ```suggestion
  to also support the unicast case that does not have a GossipSub envelope.
```

    verification procedure may pose as breaking changes and be a blocker for upgrading and keeping up with GossipSub upgrades.
4. Verification logic is not straightforward, as it requires at least two steps:
   - The first step is to ensure the event is wrapped in a GossipSub envelope. If not, the verification fails. For this we need to replicate the entire encoding path down to the GossipSub level
      as wrapping the Flow message in the GossipSub envelope is done internally at the GossipSub and is not exposed to the Flow codebase. The replication may also cause another layer of coupling 
      that causes breaking changes in the future upgrades of GossipSub.
   - The second step is to verify the signature of the GossipSub envelope against the networking public key of the origin id. This requires translating the origin id from `flow.Identifier` 
      to networking key and `peer.ID` (i.e., LibP2P level identifier), and replicating the signature verification logic of the GossipSub router, which is another layer of coupling that 
      causes breaking changes in the future upgrades of GossipSub.

> ## Proposal-2: Enforced Flow-level Signing Policy For All Messages

Reviewer 2: ```suggestion
## Proposal-2: Flow-level Signing (FLS)
```

   minimum overhead. This is a powerful feature that can be leveraged to build a robust forensic system.

### Disadvantages
1. The implementation is a **breaking change** and is not backward compatible with the current state of the Flow protocol. We change the `Conduit` interface, which is used by
   all the Flow protocol engines to send messages to the networking layer. Hence, all the Flow protocol engines must be refactored to sign the messages that are sent through
   the `Conduit` interface. 
2. This approach adds a computation overhead to the Flow protocol engines, as the engines must sign all the messages that are sent through the `Conduit` interface, and on the receiving side, the 
   networking layer must verify the signature of the message against the Staking Key of the sender. This overhead is not negligible, as the Flow protocol engines are the most performance critical components of the Flow blockchain. 
   Hence, we must carefully evaluate the performance overhead of this approach. Moreover, with this approach, we are extending the size of data sent over the wire by piggybacking the signature.

>    Assuming that we are using ECDSA with `secp521r1` curve and SHA-512 for hashing, the signature size is ~100 bytes (in theory). Hence, we are adding ~100 bytes to the size of data sent over the wire.

Reviewer 3: This example is correct, but using `secp256r1` as an example gives a closer number to what we have in Flow (`secp521` provides higher security than that we require for Flow, while `sec256r1` is what the networking layer is using). 
We can also omit the hashing algorithm because it doesn't impact the signature length. 
```suggestion
   Assuming that we are using ECDSA with `secp256r1` curve, the signature size is 64 bytes. Hence, we are adding 64 bytes to the size of data sent over the wire.
```

However, this stipulation does not apply universally across all data structures present in Flow, with `ChunkDataResponse` being a notable exception that does not demand the signature 
of its originating execution node.

Preserving the internal signature of a data structure holds considerable importance for future reference and validation processes.
This retention is especially critical when introducing data structures to new nodes, which may either be lagging or freshly joining the network; 
the internal signature by the creator facilitates the verification process, allowing the new node to establish trust in the data structure.
On the other hand, the envelope-level signature characterizes the node wrapping the data structure, diverging from the creator’s signature found in the data structure itself.
This approach substantially enhances forensic capabilities. Taking the instance of a scenario where a sender manipulates an execution receipt, compromising its integrity, 

> the system is designed to not just detect the violation via the executor's internal signature but to robustly attribute the breach to the malicious sender ensuring non-repudiation. 

Reviewer 2: ```suggestion
the system is designed to not just detect the violation via the executor's internal signature but to robustly attribute the breach to the malicious sender (via `Envelope` validation) ensuring non-repudiation.
```
